Modify the optitrack sim to take FramePoseVector

sammy-tri opened the pull request:
This is needed for MultibodyPlant support.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8998)
<!-- Reviewable:end -->

sammy-tri commented:
Tagging +(status: do not merge) until I've prepared an update for a downstream project.

---

Review status: all discussions resolved, platform LGTM missing

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LEzhZMd4NlaFqxwxd5T:b-okqnpn)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@avalenzu for feature review.  (cc @rcory as the original author and @seancurtis-tri in case my usage of `FramePoseVector` is completely bogus in some fashion)

---

Review status: all discussions resolved, platform LGTM missing, labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LEzhgLU1lQgCcznqnuL:b-gvzqlh)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Red flags started waving when I saw this PR. And a quick glance enforces it. I suspect there are problems. So, I can go over this in detail and try to itemize the problems. But, before I do, could you describe the problem you're solving/effect you're achieving?

---

Review status: all discussions resolved, LGTM missing from assignee avalenzu, platform LGTM missing, labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LEzsaN_5CkV2s3lRKs-:b-bukx0b)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
The goal is to connect the port at `MultibodyPlant::get_geometry_poses_output_port()` to `OptitrackSender` to simulate the motion capture system.  In RBP world we had a similar port which output the kinematic state of the world (see `FramePoseTracker` which got modified in this PR, that's the thing which was written for RBP).

So the idea is that any frame in your simulation can behave as if it were being tracked in a motion capture system.  The main use case historically is that you've written a planner/controller which communicates using LCM and you'd like to run the same program against your mocap-enabled real world setup and the sim.

---

Review status: all discussions resolved, LGTM missing from assignee avalenzu, platform LGTM missing, labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LEztWuM5QlAO_T2ErUb:b-gcixn1)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
FWIW I had a prior implementation where `OptitrackSender` could use either its existing input port or an input port feeding `FramePoseVector`, but it seemed like it would be simpler if I modified `FramePoseTracker` to emit the same output format as `MultibodyPlant` instead of having (effectively) a mode switch in `OptitrackSender`).

---

Review status: all discussions resolved, LGTM missing from assignee avalenzu, platform LGTM missing, labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LEzucSb8X-bTqhpqTKs:b2g84p)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I've got several concerns (and then I'll look at the PR in detail).

1. The `FrameId` values stored in the `FramePoseVector` is not particularly meaningful outside of MBP or SG. It is the identifier SG generates when MBP announces that it has a frame it's going to want to move around. That id won't be meaningful to any third part, unless MBP exposes the mapping from body id/index to FrameId.
2. MBP doesn't necessarily have a frame id for every body -- if a body doesn't have geometry, it won't have a frame id.
3. The FrameIds in the FramePoseVector won't include any arbitrary frames affixed to bodies (which I infer from your "any frame in your simulation" comment).
4. I saw a system creating an arbitrary source id. This worried me -- while technically a valid operation, it is a source id which will not be considered valid by the SG. Again, still valid, but the suggestion of SourceIds in floating around where some are valid and some aren't is...disconcerting.

So, I'll go through the PR in more detail and see if I can determine how much of this is relevant (and if there's anything I missed.)

---

Review status: all discussions resolved, LGTM missing from assignee avalenzu, platform LGTM missing, labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LF2mTetAJs_G6DYF3e3:b-frjw0o)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Ok, having read through this, I have a better sense of what's going on here. This iteration has nothing to do with MBP. If and when MBP is required, some code will build an appropriate map between `FrameId` values and name/ids to pass into the optitrack sender.

I *am* slightly uncomfortable with pulling something out of the SceneGraph echosystem and, in some sense, abusing it. In this case, I'm assuaging that discomfort by explicitly documenting an input that *looks* compatible with SceneGraph isn't, but I'm going to ponder a more robust solution. But it can be PRd later.

---

Reviewed 16 of 16 files at r1.
Review status: 4 unresolved discussions, LGTM missing from assignee avalenzu, platform LGTM missing, labeled "do not merge"

---

*[manipulation/util/frame_pose_tracker.h, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2uCF-9kesE2O0Tquj:-LF2uCF-9kesE2O0Tquk:b-c35t9a) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L21)):*
> ```objc
> 
> /**
>  * Implements a class that maintains pose and velocity information for a set of
> ```

FTR the "velocity" part is probably no longer true. And, more generally, other than the change of type, it seems like a large portion of this paragraph is no longer true.

---

*[manipulation/util/frame_pose_tracker.h, line 27 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2uPgJ0JCnB6JAK0gC:-LF2uPgJ0JCnB6JAK0gD:b-c2gcrs) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L27)):*
> ```objc
>  * existing in the RigidBodyTree), which specify which RigidBodies these newly
>  * created frames should be attached to, or (2) by providing a set of
>  * RigidBodyFrames directly.  The former simply searches the RigidBodyTree and
> ```

FYI The battle between the one-spacers and the two-spacers....ready....FIGHT! :)

---

*[manipulation/util/frame_pose_tracker.h, line 98 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF39_i94xMZfg4INkj7:-LF39_i94xMZfg4INkj8:b-ebuq0w) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L98)):*
> ```objc
>   /**
>    * This OutputPort represents an abstract valued output port of type
>    * geometry::FramePoseVector.
> ```

It should be clearly documenting that this port *cannot* be connected to SceneGraph (the attempt will throw an exception).

---

*[manipulation/util/frame_pose_tracker.h, line 101 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF390QjCsqBTerzTHNL:-LF390QjCsqBTerzTHNM:bn4tjiw) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L101)):*
> ```objc
>    */
>   const systems::OutputPort<double>& get_pose_vector_output_port() const {
>     return this->get_output_port(0);
> ```

nit: this should probably use the `pose_vector_output_port_index_` value.

---

*[manipulation/util/frame_pose_tracker.h, line 108 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2ubPp0HakzQFR707G:-LF2ubPp0HakzQFR707H:bjhdip0) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L108)):*
> ```objc
>   }
> 
>   const std::map<std::string, geometry::FrameId>&
> ```

BTW It's unfortunate that this returns a map when the name doesn't imply map. GeometryState introduces [a range iterator](https://github.com/SeanCurtis-TRI/drake/blob/master/geometry/geometry_state.h#L31) that can be used in a `for (auto frameid : foo.frame_ids())` kind of way. It may be worth pulling it into common and giving it its dual (`MapValueRange`).  Of course, it depends on whether this function is used to report size or just to iterate through the frame ids.

*OR* the name doesn't represent the return value very well.

---

*[manipulation/util/frame_pose_tracker.cc, line 9 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2zFtjBJi7ykVaizDx:-LF2zFtjBJi7ykVaizDy:bfiwikp) ([raw file](https://github.com/robotlocomotion/drake/blob/495d8aa5f89b97fa6761c58d0eaa87205246d446/manipulation/util/frame_pose_tracker.cc#L9)):*
> ```cpp
> #include "drake/multibody/multibody_tree/math/spatial_velocity.h"
> #include "drake/multibody/rigid_body_plant/kinematics_results.h"
> #include "drake/systems/rendering/frame_velocity.h"
> ```

BTW The `FrameVelocityVector` is lurking in the wings...I can PR it if you need it.

---

*[systems/sensors/optitrack_sender.cc, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32hEz2fLqpF2lB6Zq:-LF32hEz2fLqpF2lB6Zr:b-5c2tuq) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L52)):*
> ```cpp
>     const Context<double>& context,
>     optitrack::optitrack_frame_t* output) const {
>   output->utime = static_cast<int64_t >(context.get_time() * 1e6);
> ```

This "calc" method on the output port cannot assume that the `optitrack_frame_t` has the correct `num_rigid_bodies` set or a properly sized `rigid_bodies`.

---

*[systems/sensors/optitrack_sender.cc, line 55 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32v-S2SGPYEgUz8pE:-LF32v-S2SGPYEgUz8pF:b-keu8ol) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L55)):*
> ```cpp
> 
>   const geometry::FramePoseVector<double>* poses =
>       this->EvalInputValue<geometry::FramePoseVector<double>>(context, 0);
> ```

fyi: For my money, I'd like to see the `0` replaced with a name. I'd recommend a `static constexpr input_pose_port_index = 0;` kind of thing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LF2q-wb9bP58ApVHfjx:b-4j0bd9)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
-@avalenzu and +@seancurtis-tri for feature review mostly to document how it's actually gone.  :)   (Sean, thanks for jumping in here, it's much appreciated)

---

Review status: 4 unresolved discussions, LGTM missing from assignee avalenzu, platform LGTM missing, labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LF9mvp66DMG8TAZI8F-:b-927xwz)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass of feature issues addressed.  +@jwnimmer-tri for platform review under the hope that the remaining issues don't seem to point to architectural changes.

---

Review status: 3 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM missing, labeled "do not merge"

---

*[manipulation/util/frame_pose_tracker.h, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2uCF-9kesE2O0Tquj:-LFJX_L4018K-fzfFLbg:b-8o0gr5) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L21)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FTR the "velocity" part is probably no longer true. And, more generally, other than the change of type, it seems like a large portion of this paragraph is no longer true.
</blockquote></details>

Fixed the bit about velocity.  I think the rest is still accurate?  Both constructor forms still exist and should work correctly.

---

*[manipulation/util/frame_pose_tracker.h, line 27 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2uPgJ0JCnB6JAK0gC:-LFJXjep3u9i4DcxX9PB:bbr9uay) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L27)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI The battle between the one-spacers and the two-spacers....ready....FIGHT! :)
</blockquote></details>

Unfortunately, one space seems more common in this comment.  "fixed".

---

*[manipulation/util/frame_pose_tracker.h, line 98 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF39_i94xMZfg4INkj7:-LFJYkj82iBkBO30pXsS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L98)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

It should be clearly documenting that this port *cannot* be connected to SceneGraph (the attempt will throw an exception).
</blockquote></details>

Done.

---

*[manipulation/util/frame_pose_tracker.h, line 108 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2ubPp0HakzQFR707G:-LFJZKYW852kvTh_84Za:b-1te9mu) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L108)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It's unfortunate that this returns a map when the name doesn't imply map. GeometryState introduces [a range iterator](https://github.com/SeanCurtis-TRI/drake/blob/master/geometry/geometry_state.h#L31) that can be used in a `for (auto frameid : foo.frame_ids())` kind of way. It may be worth pulling it into common and giving it its dual (`MapValueRange`).  Of course, it depends on whether this function is used to report size or just to iterate through the frame ids.

*OR* the name doesn't represent the return value very well.
</blockquote></details>

Updated the name of the function.

---

*[manipulation/util/frame_pose_tracker.cc, line 9 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF2zFtjBJi7ykVaizDx:-LFJZP49CAR5t8FGHle_:b-hh24lf) ([raw file](https://github.com/robotlocomotion/drake/blob/495d8aa5f89b97fa6761c58d0eaa87205246d446/manipulation/util/frame_pose_tracker.cc#L9)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The `FrameVelocityVector` is lurking in the wings...I can PR it if you need it.
</blockquote></details>

Nothing was using the velocity information, I'm not really worried about preserving it here.

---

*[systems/sensors/optitrack_sender.cc, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32hEz2fLqpF2lB6Zq:-LFJZekk2tMDT959cfJA:b684vdc) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L52)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

This "calc" method on the output port cannot assume that the `optitrack_frame_t` has the correct `num_rigid_bodies` set or a properly sized `rigid_bodies`.
</blockquote></details>

I thought that the code in `CreateNewMessage()` above ensured exactly that?

---

*[systems/sensors/optitrack_sender.cc, line 55 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32v-S2SGPYEgUz8pE:-LFJ_IZ_AJ-vGdNi1vDc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L55)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

fyi: For my money, I'd like to see the `0` replaced with a name. I'd recommend a `static constexpr input_pose_port_index = 0;` kind of thing.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LFJa3EJC5htQ4NkKkdO:bsk8pcl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[systems/sensors/optitrack_sender.cc, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32hEz2fLqpF2lB6Zq:-LFJeyCR8UeZgPHfehnK:b-zbfmku) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L52)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I thought that the code in `CreateNewMessage()` above ensured exactly that?
</blockquote></details>

This was a common misunderstanding that we are fixing as we find them.  There is no guarantee from the framework that the `output` pointer always points to storage that has been freshly allocated (the result of a CreateNewMessage) -- any constructible value of type `optitrack_frame_t` might still be here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass done.

---

Reviewed 8 of 16 files at r1, 8 of 8 files at r2.
Review status: 9 unresolved discussions, LGTM missing from assignee jwnimmer-tri, LGTM missing from assignee seancurtis-tri, platform LGTM missing, labeled "do not merge"

---

*[examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc, line 218 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJkMyz9_x2RYKt_kER:-LFJkMyz9_x2RYKt_kES:b-anep2h) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc#L218)):*
> ```cpp
>       std::pair<std::string, int>("right_iiwa_base", 2);
>   frame_map[frame_name_to_id_map.at("box")] =
>       std::pair<std::string, int>("box", 3);
> ```

FYI I'm not sure, but can you use `{"box", 3}` here instead of explicitly calling the pair ctor?

---

*[examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc, line 224 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJkVCy3Yc1JJ8zAvJb:-LFJkVCy3Yc1JJ8zAvJc:burnsnz) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc#L224)):*
> ```cpp
>   auto optitrack_sender =
>       builder.AddSystem<OptitrackLcmFrameSender>(frame_map);
>   optitrack_sender->set_name("optitrack frame sender");
> ```

BTW System names that mimic the typename are silly and we should stop using them.  Ditto below.

---

*[manipulation/util/frame_pose_tracker.h, line 98 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJiJzY8T19m3zGDiQd:-LFJiJzY8T19m3zGDiQe:b-jn6yca) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/manipulation/util/frame_pose_tracker.h#L98)):*
> ```objc
>   /**
>    * This OutputPort represents an abstract valued output port of type
>    * geometry::FramePoseVector.  This port cannot be connect to
> ```

nit Typo `cannot be connect`

---

*[systems/sensors/optitrack_sender.h, line 14 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJcR2iAFowjmPS2sWE:-LFJcR2iAFowjmPS2sWF:b-hsrdsm) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L14)):*
> ```objc
> #include "optitrack/optitrack_frame_t.hpp"
> 
> #include "drake/common/drake_assert.h"
> ```

nit Unused by this header.

---

*[systems/sensors/optitrack_sender.h, line 29 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJcZhd11vZkPWDoRPv:-LFJcZhd11vZkPWDoRPw:b-ibwl5b) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L29)):*
> ```objc
> /// abstract-valued output port. The one abstract input port is
> /// templated on a geometry::FramePoseVector.  The abstract output
> /// port value is templated on type `optitrack_frame_t`.
> ```

FYI "templated on" isn't really the right verb for these (and below).  Maybe "is of type" or "requires" / "produces".

---

*[systems/sensors/optitrack_sender.h, line 40 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJdBEwBdda3kanb0rw:-LFJdBEwBdda3kanb0rx:b-w947so) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L40)):*
> ```objc
> 
>   /// Create an OptitrackLcmFrameSender using the FramePoseVector
>   /// input port.
> ```

nit I'm not sure what "using the FramePoseVector input port" is supposed to be telling me here.

---

*[systems/sensors/optitrack_sender.h, line 44 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJdLzgED8-gS9q-MwR:-LFJdLzgED8-gS9q-MwS:bwo0zkn) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L44)):*
> ```objc
>   /// @param frame_map contains a map of geometry::FrameId to the
>   /// rigid body name and id to be published in the resulting lcm
>   /// messages.
> ```

To a reader (me) unfamiliar with the optitrack messages, I don't grok what the string and int are -- needs more doc (or a citation to somewhere else).

Are the just the fields of https://github.com/RobotLocomotion/optitrack-driver/blob/master/lcmtypes/optitrack_rigid_body_description_t.lcm?

---

*[systems/sensors/optitrack_sender.h, line 65 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJd2ym2ofPM4hyWSjz:-LFJd2ym2ofPM4hyWSk-:b2umupg) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L65)):*
> ```objc
>       optitrack::optitrack_frame_t* output) const;
> 
>   const int num_rigid_bodies_{-1};
> ```

FYI We don't need any kind of moustache or default here, since its const.

---

*[systems/sensors/optitrack_sender.cc, line 58 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJgcI59hCW_LM9EuUs:-LFJgcI6B3MPW2fO5-0t:b-4hnojh) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.cc#L58)):*
> ```cpp
>           context, pose_input_port_index_);
> 
>   int output_idx = 0;
> ```

nit Disallowed abbreviation `idx`.

---

*[systems/sensors/optitrack_sender.cc, line 61 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJhdL-59blOS40vyC4:-LFJhdL-59blOS40vyC5:b-oy4sw2) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.cc#L61)):*
> ```cpp
>   for (const auto& frame : frame_map_) {
>     const Eigen::Isometry3d& pose = poses->value(frame.first);
>     output->rigid_bodies[output_idx].id = frame.second.second;
> ```

Unclear why `frame.second.first` (the "name") is / should be ignored.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LFJkhif0ZgPnaRBuW_7:b-te4kzy)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 5 unresolved discussions, LGTM missing from assignee jwnimmer-tri, LGTM missing from assignee seancurtis-tri, platform LGTM missing, labeled "do not merge"

---

*[examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc, line 218 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJkMyz9_x2RYKt_kER:-LFJlNzo6S2mNYWct_wF:b9bk88v) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc#L218)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I'm not sure, but can you use `{"box", 3}` here instead of explicitly calling the pair ctor?
</blockquote></details>

Done.  (much to my surprise)

---

*[examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc, line 224 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJkVCy3Yc1JJ8zAvJb:-LFJldmtBFf-1yEFeeAO:b-hkyw9l) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/examples/kuka_iiwa_arm/dev/box_rotation/iiwa_box_simulation.cc#L224)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW System names that mimic the typename are silly and we should stop using them.  Ditto below.
</blockquote></details>

Done.  (and agreed)

---

*[systems/sensors/optitrack_sender.h, line 29 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJcZhd11vZkPWDoRPv:-LFJm3eW5Y0YtL3e1j7Q:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI "templated on" isn't really the right verb for these (and below).  Maybe "is of type" or "requires" / "produces".
</blockquote></details>

Done.

---

*[systems/sensors/optitrack_sender.h, line 44 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJdLzgED8-gS9q-MwR:-LFJmacQCNHeTkhYlJuL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

To a reader (me) unfamiliar with the optitrack messages, I don't grok what the string and int are -- needs more doc (or a citation to somewhere else).

Are the just the fields of https://github.com/RobotLocomotion/optitrack-driver/blob/master/lcmtypes/optitrack_rigid_body_description_t.lcm?
</blockquote></details>

Done.

---

*[systems/sensors/optitrack_sender.h, line 65 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJd2ym2ofPM4hyWSjz:-LFJmdVR7AkIbD3t7X3k:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.h#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI We don't need any kind of moustache or default here, since its const.
</blockquote></details>

Done.

---

*[systems/sensors/optitrack_sender.cc, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32hEz2fLqpF2lB6Zq:-LFJn68908CrZR5PNqVx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This was a common misunderstanding that we are fixing as we find them.  There is no guarantee from the framework that the `output` pointer always points to storage that has been freshly allocated (the result of a CreateNewMessage) -- any constructible value of type `optitrack_frame_t` might still be here.
</blockquote></details>

Done.

---

*[systems/sensors/optitrack_sender.cc, line 58 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJgcI59hCW_LM9EuUs:-LFJnJC12NIFDjicaxYp:b-urk1hn) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.cc#L58)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Disallowed abbreviation `idx`.
</blockquote></details>

Done, though I'd argue that counts as a widespread convention like `num` if I really cared (and I do not).

---

*[systems/sensors/optitrack_sender.cc, line 61 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFJhdL-59blOS40vyC4:-LFJnRwkAgUD9ZzOftX8:bg34j2k) ([raw file](https://github.com/robotlocomotion/drake/blob/02cc454b7e29db2ab556b6098ee9790d1da6ef52/systems/sensors/optitrack_sender.cc#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why `frame.second.first` (the "name") is / should be ignored.
</blockquote></details>

I updated the comments in the header to attempt to indicate that this is related to the TODO for publishing the description messages.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> platform.

---

Reviewed 4 of 4 files at r3.
Review status: 3 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM from [jwnimmer-tri], labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LFJpvJvDUIoTp-F-azB:b-vtj4vo)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Dismissed @SeanCurtis-TRI from a discussion.
Review status: 2 unresolved discussions, LGTM missing from assignee seancurtis-tri, platform LGTM from [jwnimmer-tri], labeled "do not merge"

---

*[systems/sensors/optitrack_sender.cc, line 52 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF32hEz2fLqpF2lB6Zq:-LFJrXtJ6R9gx94av7Wv:b5rkxyv) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/systems/sensors/optitrack_sender.cc#L52)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Dismissing since Jeremy set me straight here and the issue seems fixed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 8 files at r2, 4 of 4 files at r3.
Review status: 1 unresolved discussion, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---

*[manipulation/util/frame_pose_tracker.h, line 98 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF39_i94xMZfg4INkj7:-LFLBbn_BXCAaVPARWH3:b-sp57tw) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L98)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

FYI As to the earlier note: you've got more two-space sentences here as well.

---

*[systems/sensors/optitrack_sender.h, line 40 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFLBud81msNHU-9IZAQ:-LFLBud81msNHU-9IZAR:b-2f9h5d) ([raw file](https://github.com/robotlocomotion/drake/blob/f79fdbcfb5d96a87eb74c81a2a19c9f9ba16523b/systems/sensors/optitrack_sender.h#L40)):*
> ```objc
>   /// name and id to be published in the resulting `optitrack_rigid_body_t` and
>   /// `optitrack_rigid_body_description_t` messages (descriptions not yet
>   /// implemented)..
> ```

FYI Two periods

---

*[systems/sensors/optitrack_sender.cc, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFLC39O6MvStWornORI:-LFLC39O6MvStWornORJ:b8v8ypx) ([raw file](https://github.com/robotlocomotion/drake/blob/f79fdbcfb5d96a87eb74c81a2a19c9f9ba16523b/systems/sensors/optitrack_sender.cc#L45)):*
> ```cpp
>   output->utime = static_cast<int64_t >(context.get_time() * 1e6);
>   output->num_rigid_bodies = num_rigid_bodies_;
>   output->rigid_bodies.resize(static_cast<size_t>(num_rigid_bodies_));
> ```

FYI For my own edification, in this case, I tend to go fore a `reserve` in place of a `resize`. My understanding is that the resize will actually do writing to the allocated memory, whereas `reserve` doesn't. Both serve the purpose of making a single call to memory allocation (as necessary). I'm curious to hear other opinions on the subject. Definitely not blocking.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LFLCN2f-2ut1QkZ6yfE:bmxilld)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 1 unresolved discussion, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---

*[systems/sensors/optitrack_sender.cc, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFLC39O6MvStWornORI:-LFMvP6C6oU7WlElUcoG:b7fkd1r) ([raw file](https://github.com/robotlocomotion/drake/blob/f79fdbcfb5d96a87eb74c81a2a19c9f9ba16523b/systems/sensors/optitrack_sender.cc#L45)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI For my own edification, in this case, I tend to go fore a `reserve` in place of a `resize`. My understanding is that the resize will actually do writing to the allocated memory, whereas `reserve` doesn't. Both serve the purpose of making a single call to memory allocation (as necessary). I'm curious to hear other opinions on the subject. Definitely not blocking.
</blockquote></details>

Its not even about the memory stuff -- there is no meaningful performance difference in this case.  It's about whether, when filling this in later, we'd prefer to use assignment (resize) or emplacement (reserve).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: 1 unresolved discussion, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---

*[systems/sensors/optitrack_sender.cc, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFLC39O6MvStWornORI:-LFNKqZq7qch5h6gOTsa:b2mowdo) ([raw file](https://github.com/robotlocomotion/drake/blob/f79fdbcfb5d96a87eb74c81a2a19c9f9ba16523b/systems/sensors/optitrack_sender.cc#L45)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Its not even about the memory stuff -- there is no meaningful performance difference in this case.  It's about whether, when filling this in later, we'd prefer to use assignment (resize) or emplacement (reserve).
</blockquote></details>

In this case, true; it's a bunch of `size_t`. But I think of it like doing a `++itr` instead of `itr++`. For many common cases, the difference is somewhere between negligible and non-existent. But sometimes it does matter. So, I've gotten into the habit of using the more generally optimal pattern rather than curating it based on the underlying type.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 1 unresolved discussion, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---

*[systems/sensors/optitrack_sender.cc, line 45 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFLC39O6MvStWornORI:-LFNOU9S8_0RsW3IUQUj:b-frq4z4) ([raw file](https://github.com/robotlocomotion/drake/blob/f79fdbcfb5d96a87eb74c81a2a19c9f9ba16523b/systems/sensors/optitrack_sender.cc#L45)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

In this case, true; it's a bunch of `size_t`. But I think of it like doing a `++itr` instead of `itr++`. For many common cases, the difference is somewhere between negligible and non-existent. But sometimes it does matter. So, I've gotten into the habit of using the more generally optimal pattern rather than curating it based on the underlying type.
</blockquote></details>

(1) The contained type is not `size_t`, its a struct.
(2) That's still off-topic.  If you want to use assignment in the loop, you use resize to prep for it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 1 unresolved discussion, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---

*[manipulation/util/frame_pose_tracker.h, line 98 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LF39_i94xMZfg4INkj7:-LFNYMP-8KARYjOHC5Fu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e4b358696e2c54e93cd7099d1e77819c708a9dd2/manipulation/util/frame_pose_tracker.h#L98)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI As to the earlier note: you've got more two-space sentences here as well.
</blockquote></details>

Done.

---

*[systems/sensors/optitrack_sender.h, line 40 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8998#-LFLBud81msNHU-9IZAQ:-LFNY-QL8kBagdW03LWC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f79fdbcfb5d96a87eb74c81a2a19c9f9ba16523b/systems/sensors/optitrack_sender.h#L40)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Two periods
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r4.
Review status: 1 unresolved discussion, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Dismissed @SeanCurtis-TRI from a discussion.
Review status: all discussions resolved, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
This is ready, other than the DNM tag.

sammy-tri commented:
-(status: do not merge) 

---

Review status: all discussions resolved, platform LGTM from [jwnimmer-tri, seancurtis-tri], labeled "do not merge"

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8998#-:-LFOKmlFD85gm6ZYSe-A:b-b9fcpn)*
<!-- Sent from Reviewable.io -->
